File: web/src/features/admin-support/api/refunds.ts
/**
 * Admin 수동 환불 검토 API 클라이언트 — Story 9.3.
 *
 *   GET   /api/v1/admin/refunds                       (status/from/to/q + 페이지네이션)
 *   POST  /api/v1/admin/refunds/{queue_id}/approve    (note 필수)
 *   POST  /api/v1/admin/refunds/{queue_id}/deny       (note 필수)
 *
 * 모든 mutate 호출에 X-CSRF-Token 헤더 부착.
 */

const API_BASE = process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8000";

export type RefundQueueStatus = "pending" | "approved" | "denied";
export type RefundReasonCode =
  | "period_exceeded"
  | "qa_count_exceeded"
  | "both"
  | "no_subscription";

export interface RefundQueueItem {
  queue_id: number;
  payment_id: number;
  user_id: number;
  user_email_masked: string;
  amount_krw: number;
  card_last4: string | null;
  card_company: string | null;
  days_since_charge: number;
  qa_count_during_period: number;
  reason_code: RefundReasonCode | null;
  reason_text: string | null;
  requested_at: string;
  status: RefundQueueStatus;
  reviewer_user_email_masked: string | null;
  reviewer_note: string | null;
  reviewed_at: string | null;
}

export interface RefundQueueListResponse {
  items: RefundQueueItem[];
  page: number;
  per_page: number;
  total: number;
}

export interface FetchRefundsParams {
  status?: RefundQueueStatus;
  from?: string;
  to?: string;
  q?: string;
  page?: number;
  per_page?: number;
}

export interface RefundActionResponse {
  queue_id: number;
  payment_id: number;
  status: RefundQueueStatus;
  amount_krw: number;
  refunded_at: string | null;
}

export class RefundActionError extends Error {
  status: number;
  code: string;
  traceId?: string;
  details?: Record<string, unknown>;

  constructor(
    status: number,
    code: string,
    message: string,
    traceId?: string,
    details?: Record<string, unknown>,
  ) {
    super(message);
    this.name = "RefundActionError";
    this.status = status;
    this.code = code;
    this.traceId = traceId;
    this.details = details;
  }
}

function _readCookie(name: string): string | undefined {
  if (typeof document === "undefined") return undefined;
  const match = document.cookie.match(
    new RegExp("(?:^|; )" + name.replace(/[-]/g, "\\$&") + "=([^;]*)"),
  );
  return match ? decodeURIComponent(match[1]) : undefined;
}

function _withCsrf(headers: Record<string, string>): Record<string, string> {
  const csrf = _readCookie("denvia_admin_csrf");
  if (csrf) headers["X-CSRF-Token"] = csrf;
  return headers;
}

async function _toError(res: Response, fallback: string): Promise<RefundActionError> {
  let code = "UNKNOWN_ERROR";
  let message = fallback;
  let traceId: string | undefined;
  let details: Record<string, unknown> | undefined;
  try {
    const body = (await res.json()) as {
      code?: string;
      message?: string;
      trace_id?: string;
      details?: Record<string, unknown>;
    };
    if (body.code) code = body.code;
    if (body.message) message = body.message;
    if (body.trace_id) traceId = body.trace_id;
    if (body.details) details = body.details;
  } catch {
    /* fallthrough */
  }
  return new RefundActionError(res.status, code, message, traceId, details);
}

export async function fetchRefundQueue(
  params: FetchRefundsParams = {},
): Promise<RefundQueueListResponse> {
  const query = new URLSearchParams();
  if (params.status) query.set("status", params.status);
  if (params.from) query.set("from", params.from);
  if (params.to) query.set("to", params.to);
  if (params.q) query.set("q", params.q);
  if (params.page) query.set("page", String(params.page));
  if (params.per_page) query.set("per_page", String(params.per_page));

  const res = await fetch(
    `${API_BASE}/api/v1/admin/refunds?${query.toString()}`,
    { credentials: "include" },
  );
  if (!res.ok) {
    throw new Error(`admin refunds fetch failed: ${res.status}`);
  }
  return res.json() as Promise<RefundQueueListResponse>;
}

export async function approveRefund(
  queueId: number,
  note: string,
): Promise<RefundActionResponse> {
  const headers = _withCsrf({ "Content-Type": "application/json" });
  const res = await fetch(
    `${API_BASE}/api/v1/admin/refunds/${queueId}/approve`,
    {
      method: "POST",
      credentials: "include",
      headers,
      body: JSON.stringify({ note }),
    },
  );
  if (!res.ok) {
    throw await _toError(res, "환불 승인에 실패했습니다.");
  }
  return res.json() as Promise<RefundActionResponse>;
}

export async function denyRefund(
  queueId: number,
  note: string,
): Promise<RefundActionResponse> {
  const headers = _withCsrf({ "Content-Type": "application/json" });
  const res = await fetch(
    `${API_BASE}/api/v1/admin/refunds/${queueId}/deny`,
    {
      method: "POST",
      credentials: "include",
      headers,
      body: JSON.stringify({ note }),
    },
  );
  if (!res.ok) {
    throw await _toError(res, "환불 거부에 실패했습니다.");
  }
  return res.json() as Promise<RefundActionResponse>;
}

File: api/src/schemas/admin/refunds.py
"""Admin 수동 환불 검토 — Pydantic 스키마 (Story 9.3 A-503).

GET   /api/v1/admin/refunds                       manual_refund_queue 목록 (status/from/to/q + 페이지네이션)
POST  /api/v1/admin/refunds/{queue_id}/approve    PG refund 호출 + payment.status='refunded' + 알림톡 + 쪽지함
POST  /api/v1/admin/refunds/{queue_id}/deny       payment.status 원복 + 알림톡(거부) + 쪽지함

설계 원칙:
- 응답은 admin/users/finance 와 동일한 flat 페이지네이션(items/page/per_page/total).
- approve/deny 모두 reviewer_note 필수 (감사 추적).
- 알림톡 발송은 fire-and-forget — 응답 성공 여부에 영향 없음 (3.6 _execute_auto_refund 패턴 답습).
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

RefundQueueStatus = Literal["pending", "approved", "denied"]
RefundReasonCode = Literal[
    "period_exceeded",
    "qa_count_exceeded",
    "both",
    "no_subscription",
]


class RefundQueueItem(BaseModel):
    """환불 큐 1행."""

    queue_id: int
    payment_id: int
    user_id: int
    user_email_masked: str
    amount_krw: int
    card_last4: str | None = None
    card_company: str | None = None
    days_since_charge: int
    qa_count_during_period: int
    reason_code: RefundReasonCode | None = None
    reason_text: str | None = Field(
        default=None,
        description="manual_refund_queue.reason — 사용자 입력 사유 (null 가능)",
    )
    requested_at: datetime
    status: RefundQueueStatus
    reviewer_user_email_masked: str | None = None
    reviewer_note: str | None = None
    reviewed_at: datetime | None = None


class RefundQueueListResponse(BaseModel):
    """GET /api/v1/admin/refunds — flat 페이지네이션."""

    items: list[RefundQueueItem]
    page: int
    per_page: int
    total: int


class RefundActionRequest(BaseModel):
    """POST /api/v1/admin/refunds/{queue_id}/approve|deny — 검토자 메모 필수."""

    note: str = Field(
        min_length=1,
        max_length=1000,
        description="검토자 메모. 거부 시 사용자에게 일부(80자)가 알림톡으로 노출됨.",
    )


class RefundActionResponse(BaseModel):
    """approve/deny 응답."""

    queue_id: int
    payment_id: int
    status: RefundQueueStatus
    amount_krw: int
    refunded_at: datetime | None = Field(
        default=None,
        description="approve 시점 refund 완료 시각. deny 시 None.",
    )


__all__ = [
    "RefundQueueStatus",
    "RefundReasonCode",
    "RefundQueueItem",
    "RefundQueueListResponse",
    "RefundActionRequest",
    "RefundActionResponse",
]
